### PR TITLE
Allow for access of final progress stats

### PR DESCRIPTION
**Pull request by alex-hofsteede**
This refactors some of the logic around the processing of progress
packets so that the final (sum) of progress counters is accessible. It
includes the following changes:

- make QueryResult know how to process and `store()` its own packets.
- Make ProgressQueryResult class a subclass of QueryResult.
- ProgressQueryResult can process progress packets.
- Less encapsulation breakage (Client reaching into QueryResult objects)
- Client now provides generator of packets to QueryResult to iterate on
- ProgressQueryResult now aggregates bytes as well as rows (although
bytes are not returned in the generator to preserve the current
interface)
- Final sum of progress counters is accessible as `progress_totals`

### CODE REVIEW

File: src/client.py [hunk 1]
```diff
@@ -1,45 +1,83 @@
...
 
     def next(self):
-        try:
-            return next(self.progress_gen)
-        except Exception as ex:
-            if not isinstance(ex, StopIteration):
-                self.connection.disconnect()
-            raise
+        while True:
+            packet = next(self.packet_generator)
```
Comment (xzkostyan) [inline src/client.py]: We should `.disconnect()` if non-`StopIteration` exception raises and re-raise this exception. See previous code. This in the `packet_generator` now. There is no need in disconnect on `StopIteration`

File: src/client.py [hunk 2]
```diff
@@ -1,45 +1,83 @@
...
         else:
             return self.data
 
 
-class Progress(object):
-    def __init__(self, client_connection, result, progress_gen):
-        self.connection = client_connection
-        self.result = result
-        self.progress_gen = progress_gen
+class ProgressQueryResult(QueryResult):
```
Comment (xzkostyan) [inline src/client.py]: Let's refactor a little bit and place `ProgressQueryResult` and `QueryResult` in `result.py` file.

Comment (alex-hofsteede) [reply src/client.py]: Good idea, done

File: src/client.py (hunk 1 above)
Comment (alex-hofsteede) [reply src/client.py]: Not 100% sure what you mean here. By moving the exception handler with the `disconnect()` call down to `packet_generator`, I avoided the need to add a case for `StopIteration`, as the `StopIteration` would now only be raised in `ProgressQueryResult.next()`, where we do not need to know about the connection or disconnect it. 

I _think_ the overall behavior is the same as it used to be, but I could be missing something.

**Pull request merged**